### PR TITLE
[CLOUDSTACK-10356] Fix NPE in Cloudstack found with NPEDetector 

### PR DESCRIPTION
**Pull request by brettKK**
@lujiefsi found a potential NPE in Cloudstack
---
We have developed a static analysis tool [NPEDetector](https://github.com/lujiefsi/NPEDetector) to find some potential NPE. Our analysis shows that some callees may return null in corner case(e.g. node crash , IO exception), some of their callers have  !=null check but some do not have. In this issue we post a patch which can add  !=null  based on existed !=null  check. For example:

Callee GlobalLoadBalancingRulesServiceImpl#lookupGslbServiceProvider:
```
protected GslbServiceProvider lookupGslbServiceProvider() {
    return _gslbProviders.size() == 0 ? null : _gslbProviders.get(0);// may return null;
}
```

Caller GlobalLoadBalancingRulesServiceImpl#checkGslbServiceEnabledInZone have !=null:
```
private boolean checkGslbServiceEnabledInZone(long zoneId, long physicalNetworkId) {

   GslbServiceProvider gslbProvider = lookupGslbServiceProvider();
   if (gslbProvider == null) {
      throw new CloudRuntimeException("No GSLB provider is available");
   }

   return gslbProvider.isServiceEnabledInZone(zoneId, physicalNetworkId);
}
```

but another GlobalLoadBalancingRulesServiceImpl#applyGlobalLoadBalancerRuleConfig does not have !=null check:
```
GslbServiceProvider gslbProvider = lookupGslbServiceProvider();
siteLb.setGslbProviderPublicIp(gslbProvider.getZoneGslbProviderPublicIp(dataCenterId,physicalNetworkId));
.........
```

So we will add below code in non-(!=null) caller GlobalLoadBalancingRulesServiceImpl#applyGlobalLoadBalancerRuleConfig
```
if (gslbProvider == null) {
    throw new CloudRuntimeException("No GSLB provider is available");
}
```

But due to we are not very  familiar with CLOUDSTACK, hope some expert can review it.

Thanks.

### CODE REVIEW

**Comment by brettKK** [comment]
fix describe

**Comment by DaanHoogland** [comment]
removed a few tab chars @brettKK , sorry to not PR them but github failed at that

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 1]
```diff
@@ -499,6 +499,9 @@ public VolumeInfo copyVolumeFromSecToPrimary(VolumeInfo volume, VirtualMachine v
 
         // Find a suitable storage to create volume on
         StoragePool destPool = findStoragePool(dskCh, dc, pod, clusterId, null, vm, avoidPools);
+        if (destPool == null) {
+            throw new CloudRuntimeException("Failed to find a storage pool with enough capacity to copy the volume to.");
```
Comment (rafaelweingartner) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java]: This message looks misleading. Will tags be considered at this point as well?

File: engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java [hunk 2]
```diff
@@ -126,8 +126,10 @@ public void updateDomainCount(long domainId, ResourceType type, boolean incremen
         delta = increment ? delta : delta * -1;
 
         ResourceCountVO resourceCountVO = findByOwnerAndType(domainId, ResourceOwnerType.Domain, type);
-        resourceCountVO.setCount(resourceCountVO.getCount() + delta);
-        update(resourceCountVO.getId(), resourceCountVO);
+        if (resourceCountVO != null) {
```
Comment (rafaelweingartner) [inline engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java]: If you do not find a resource to update, should not you break the execution flow?
It seems that if the `resourceCountVO` is null, then we must have something wrong in the database.

File: plugins/deployment-planners/implicit-dedication/src/main/java/com/cloud/deploy/ImplicitDedicationPlanner.java [hunk 3]
```diff
@@ -256,14 +256,15 @@ public PlannerResourceUsage getResourceUsage(VirtualMachineProfile vmProfile, De
 
             // Get the list of all the hosts in the given clusters
             List<Long> allHosts = new ArrayList<Long>();
-            for (Long cluster : clusterList) {
-                List<HostVO> hostsInCluster = resourceMgr.listAllHostsInCluster(cluster);
-                for (HostVO hostVO : hostsInCluster) {
+            if (clusterList != null && !clusterList.isEmpty()) {
```
Comment (rafaelweingartner) [inline plugins/deployment-planners/implicit-dedication/src/main/java/com/cloud/deploy/ImplicitDedicationPlanner.java]: You can use `CollectionUtils.isNotEmpty` here.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 4]
```diff
@@ -2337,8 +2337,9 @@ public int compare(final DiskTO arg0, final DiskTO arg1) {
                     disk.setCacheMode(DiskDef.DiskCacheMode.valueOf(volumeObjectTO.getCacheMode().toString().toUpperCase()));
                 }
             }
-
-            vm.getDevices().addDevice(disk);
+            if (vm.getDevices() != null) {
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Should not an exception be thrown in this case?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 5]
```diff
@@ -2391,8 +2392,10 @@ private void createVif(final LibvirtVMDef vm, final NicTO nic, final String nicA
                         + ") is " + nic.getType() + " traffic type. So, vsp-vr-ip " + vrIp + " is set in the metadata");
             }
         }
-
-        vm.getDevices().addDevice(getVifDriver(nic.getType(), nic.getName()).plug(nic, vm.getPlatformEmulator(), nicAdapter));
+        
+        if (vm.getDevices() != null) {
```
Comment (rafaelweingartner) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: The same here?

File: server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java [hunk 6]
```diff
@@ -2139,14 +2139,14 @@ public boolean startRemoteAccessVpn(final Network network, final RemoteAccessVpn
             }
 
             Answer answer = cmds.getAnswer("users");
-            if (!answer.getResult()) {
+            if (answer != null && !answer.getResult()) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java]: You can use `org.apache.commons.lang3.BooleanUtils.toBooleanObject(String, String, String, String)` here

File: server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java [hunk 7]
```diff
@@ -2139,14 +2139,14 @@ public boolean startRemoteAccessVpn(final Network network, final RemoteAccessVpn
...
                 s_logger.error("Unable to start vpn: unable add users to vpn in zone " + router.getDataCenterId() + " for account " + vpn.getAccountId() + " on domR: "
                         + router.getInstanceName() + " due to " + answer.getDetails());
                 throw new ResourceUnavailableException("Unable to start vpn: Unable to add users to vpn in zone " + router.getDataCenterId() + " for account "
                         + vpn.getAccountId() + " on domR: " + router.getInstanceName() + " due to " + answer.getDetails(), DataCenter.class, router.getDataCenterId());
             }
             answer = cmds.getAnswer("startVpn");
-            if (!answer.getResult()) {
+            if (answer != null && !answer.getResult()) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java]: Same here

File: server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java [hunk 8]
```diff
@@ -740,14 +740,14 @@ public boolean startRemoteAccessVpn(final RemoteAccessVpn vpn, final VirtualRout
             throw new AgentUnavailableException("Unable to send commands to virtual router ", router.getHostId(), e);
         }
         Answer answer = cmds.getAnswer("users");
-        if (!answer.getResult()) {
+        if (answer != null && !answer.getResult()) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java]: Here as well

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java (hunk 1 above)
Comment (lujiefsi) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java]: how about: throw new CloudRuntimeException("Failed to find a suitable storage pool to create Volume in the pod/cluster of the provided VM "+ vm.getUuid());

File: engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java (hunk 2 above)
Comment (lujiefsi) [reply engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java]: updateDomainCount is an @Deprecated method, can we ignore it?

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java (hunk 1 above)
Comment (rafaelweingartner) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java]: I think this one that you suggested is better.

File: engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java (hunk 2 above)
Comment (rafaelweingartner) [reply engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java]: I did not check the code. Is it used somewhere else?
I think we can ignore (it has been working so far...).

Comment (rafaelweingartner) [reply engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java]: Then, what about if you delete it?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java (hunk 4 above)
Comment (rafaelweingartner) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: I only mentioned an exception for a single reason. Right now the code works; and, if we ever get a null calling `vm.getDevices() `, a run time exception will be thrown, and we will be able to easily pin point the problem. However, after applying your code, for sure something else will break or not work as expected, which will make more difficult for us to pin point the problem later (if it ever happens).

Comment (lujiefsi) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: hum. I mean another three "vm.getDevices;" are all in the same function,  vm.getDevices can always return same object, so add checker at  line 2340 mean that check the parameter vm is legal.

Comment (rafaelweingartner) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Actually, I am not talking about the variable `vm`. I am instead, referring to the return of `vm.getDevices()` method.

File: engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java (hunk 2 above)
Comment (lujiefsi) [reply engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java]: hum, you mean delete the updateDomainCount ?

Comment (rafaelweingartner) [reply engine/schema/src/main/java/com/cloud/configuration/dao/ResourceCountDaoImpl.java]: Yes. If it is not used. So, it is better to delete it.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java (hunk 4 above)
Comment (lujiefsi) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: i got, the exception stack can help us where error happens accurately .

File: server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rafaelweingartner) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java]: Sorry, I did not understand your question. Are you asking if I would suggest to use BooleanUtils here as well? Please, go ahead and do that.

**Comment by DaanHoogland** [comment]
@brettKK this can not be merged, I think you based on master and then pointed the pull request at 4.11. I think you need to rebase your changes on the branch you want to merge to.

**Comment by rafaelweingartner** [comment]
@brettKK you can point it directly to master if you want. Then, you do not need any re-work.
If you want to target 4.11 instead of master, you indeed need to reset your branch to 4.11, and then apply your changes.

**Comment by DaanHoogland** [comment]
@brettKK what @rafaelweingartner is refering to is the following procedure:
$ git checkout -b 4.11 origin/4.11
$ git pull # may not be needed anymore ;)
$ git checkout -b CLOUDSTACK-10356-4.11
$ git cherry-pick dc8f4ef
# other cherry picks in order of your choice but make sure you have them all
$ git cherry-pick 2856781 
# alternatively you can squash all changes first and do a siingle cherry-pick
# google for a howto
# and then
$ git branch -m CLOUDSTACK-10356 CLOUDSTACK-10356-master
$ git branch -m CLOUDSTACK-10356-4.11 CLOUDSTACK-10356
$ git push --force brettkk CLOUDSTACK-10356 # or whatever alias you use for your own fork.

**Comment by DaanHoogland** [comment]
i reset the base to master in case someone accidentally would merge

File: server/src/main/java/com/cloud/projects/ProjectManagerImpl.java [hunk 9]
```diff
@@ -479,7 +479,7 @@ public void doInTransactionWithoutResult(TransactionStatus status) throws Resour
                 throw new InvalidParameterValueException("Unable to find account name=" + newOwnerName + " in domain id=" + project.getDomainId());
             }
             Account currentOwnerAccount = getProjectOwner(projectId);
-            if (currentOwnerAccount.getId() != futureOwnerAccount.getId()) {
+            if (currentOwnerAccount !=null && currentOwnerAccount.getId() != futureOwnerAccount.getId()) {
                 ProjectAccountVO futureOwner = _projectAccountDao.findByProjectIdAccountId(projectId, futureOwnerAccount.getAccountId());
```
Comment (lujiefsi) [inline server/src/main/java/com/cloud/projects/ProjectManagerImpl.java]: Here, I think when currentOwnerAccount ==null, we should throw InvalidParameterValueException.

**Comment by lujiefsi** [comment]
Seem that this pull request does not associate with JIRA  [CLOUDSTACK-10356](https://issues.apache.org/jira/browse/CLOUDSTACK-10356)

**Comment by lujiefsi** [comment]
We should also need to delete the method updateDomainCount in ResourceCountDao, which is also @Deprecated

**Comment by DaanHoogland** [comment]
so @brettKK @lujiefsi @rafaelweingartner (@rhtyd @swill @wido ) and of course all others
are we waiting for this to be rebased against 4.11 or do we merge to master?

**Comment by lujiefsi** [comment]
merge to master is better?

**Comment by brettKK** [comment]
@DaanHoogland , sry, I couldn't get your mean.  we can't merge CLOUDSTACK-10356 to apache:master. Only the some authoried guys can do that operation

**Comment by DaanHoogland** [comment]
no @lujiefsi , just less work for @brettKK . It will not be available in 4.11.1 that way, though

**Comment by rafaelweingartner** [comment]
It is the contributor choice the branch he/she is sending the contributions. If you already did everything in master, there is no reason to use even more time to port to 4.11. Unless you need these changes in 4.11.

Moreover, for us to be able to execute the merge, we need reviewers to approve and integration tests to pass. As soon as we have the test results, and if they are positive, we can proceed and merge this PR.

**Comment by rhtyd** [comment]
Tests LGTM, most changes are null checks. I'll kick test.
@blueorangutan test

**Comment by lujiefsi** [comment]
Are the errors caused by the patch? or  can the patch be merged?

**Comment by rafaelweingartner** [comment]
@lujiefsi no, they do not seem to be. However, to be safe, can @brettKK download the log files and check the errors/failures there? Just to make sure that the errors are not related to the changes introduced here.

**Comment by lujiefsi** [comment]
@rafaelweingartner I extract the exception from the log, and check the code around them, those exception seems not caused by this  patch. @brettKK could you please check again.

**Comment by rafaelweingartner** [comment]
that is a good thing. Sometimes this happens. The errors may be due to environmental problems in the test bed.

**Comment by brettKK** [comment]
@rafaelweingartner @lujiefsi I locate all exception location from the logs:

CloudStackPrimaryDataStoreLifeCycleImpl#319 
UserVmManagerImpl#5044
DeployVMCmd#567
ResourceManagerImpl#1278
AssertionError: Check uptime is less than 3 mins or not
AssertionError: Check for warnings in tests

these exception and their code are really not related to our code.

**Comment by rhtyd** [comment]
@blueorangutan package

**Pull request merged**